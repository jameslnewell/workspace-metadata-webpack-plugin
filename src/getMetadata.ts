import * as path from 'path';
import * as glob from 'fast-glob';
import findAndReadPackageJSON from 'find-and-read-package-json';

export interface Package {
  path: string;
  json: {[name: string]: any};
}

export interface Metadata {
  root: Package;
  workspaces: Package[];
}

async function getRootMetadata(cwd: string): Promise<Package> {
  const {file, json} = await findAndReadPackageJSON(cwd);
  return {
    path: path.dirname(file),
    json
  };
}

async function getWorkspaceMetadata(root: Package): Promise<Package[]> {
  const patterns = root.json.workspaces;

  if (!Array.isArray(patterns)) {
    return [];
  }

  const files = await glob<string>(
    patterns.map(pattern => `${pattern}/package.json`),
    {
      cwd: root.path,
      absolute: true
    }
  );

  return files.map(file => {
    return {
      path: path.dirname(file),
      json: require(file)
    }
  });
}

export async function getMetadata(cwd: string): Promise<Metadata> {
  const root = await getRootMetadata(cwd);
  const workspaces = await getWorkspaceMetadata(root);
  return {
    root,
    workspaces
  };
}
