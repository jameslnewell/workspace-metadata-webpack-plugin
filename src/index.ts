export * from './getMetadata';
export * from './plugin';
