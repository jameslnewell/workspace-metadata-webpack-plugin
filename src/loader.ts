import * as webpack from 'webpack';
import * as loaderUtils from 'loader-utils';
import * as validateOptions from 'schema-utils';
import { Metadata, getMetadata } from './getMetadata';

const schema = {
  type: 'object',
  properties: {
    context: {
      type: 'string'
    },
    transform: {
      instanceof: 'Function'
    },
    stringify: {
      instanceof: 'Function'
    }
  }
};

function defaultTransform<M>(metadata: Metadata) {
  return metadata as any;
}

function defaultStringify<M>(metadata: M) {
  return `export default ${JSON.stringify(metadata)}`;
}

function getDependencies(metadata: Metadata) {
  const {root, workspaces} = metadata;
  return [
    `${root.path}/package.json`,
    ...workspaces.map(workspace => `${workspace.path}/package.json`)
  ];
}

export default async function(this: webpack.loader.LoaderContext, source: string) {
  const callback = this.async();
  try {

    // check the options
    const options = loaderUtils.getOptions(this) || {};
    validateOptions(schema, options, 'workspace-metadata loader');

    // get the default options
    const {
      context = this.rootContext, 
      transform = defaultTransform, 
      stringify = defaultStringify
    } = options;

    // get the metadata
    let metadata = await getMetadata(context);
    
    // change paths to relative ones since absolute ones break the hashing - use `stringifyRequest`
    // @see https://webpack.js.org/contribute/writing-a-loader/#absolute-paths
    metadata.root.path = loaderUtils.stringifyRequest(this, metadata.root.path);
    metadata.workspaces = metadata.workspaces.map(workspace => ({
      ...workspace,
      path: loaderUtils.stringifyRequest(this, workspace.path)
    }));

    // the output can be cached
    this.cacheable(true);

    // list any files we use so that webpack rebundles when they change
    getDependencies(metadata).forEach(dependency => {
      this.addDependency(dependency);
    });

    // transform the metadata
    if (transform) {
      metadata = await transform.call(this, metadata);
    }

    // convert the metadata to a string
    const output = await stringify.call(this, metadata);

    if (callback) {
      callback(null, output);
    }
  } catch (error) {
    if (callback) {
      callback(error);
    }
  }
}
