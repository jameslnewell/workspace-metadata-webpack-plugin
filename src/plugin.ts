import * as path from 'path';
import * as webpack from 'webpack';
import { Metadata } from './getMetadata';
import { Compiler } from 'webpack';

// fix for webpack v4
declare module "webpack" {
  namespace loader {
    interface LoaderContext {
      rootContext: string;
    }
  }
}

export interface PluginOptions<M> {
  context?: string;
  module?: string;
  transform?: (this: webpack.loader.LoaderContext, metadata: Metadata) => Promise<M> | M;
  stringify?: (metadata: M) => Promise<string> | string;
}

export class Plugin<M = Metadata> {

  constructor(private options: PluginOptions<M> = {}) {
  }

  apply(compiler: webpack.Compiler) {
    const {
      context,
      module,
      transform,
      stringify
    } = this.options;

    // add the loader
    compiler.options.module = {
      ...compiler.options.module,
      rules: [
        ...(compiler.options.module ? compiler.options.module.rules : []),
        {
          test: /workspace-metadata-loader.txt$/,
          include: path.dirname(__dirname),
          loader: require.resolve('./loader'),
          options: {
            context,
            transform,
            stringify
          }
        }
      ]
    }
    
    // make a nice alias for the loader
    compiler.options.resolve = {
      ...compiler.options.resolve,
      alias: {
        ...(compiler.options.resolve ? compiler.options.resolve.alias : {}),
        [module || 'workspace-metadata']: require.resolve('./workspace-metadata-loader.txt')
      }
    }

  }

}