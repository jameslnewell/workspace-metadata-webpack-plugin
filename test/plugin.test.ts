import * as path from 'path';
import * as webpack from 'webpack';
import MemoryFS = require('memory-fs');
import {Plugin} from '../src';
import {project2Path, project3Path} from './fixtures';

// fix for webpack v4
declare module "webpack" {
  namespace loader {
    interface LoaderContext {
      rootContext: string;
    }
  }
}

function run(options: webpack.Configuration): Promise<webpack.Stats> {
  return new Promise((resolve, reject) => {
    const compiler = webpack(options);
    compiler.outputFileSystem = new MemoryFS() as any;
    compiler.run((error, stats) => {
      if (error) {
        reject(error);
      } else {
        expect(stats.compilation.errors).toHaveLength(0);
        expect(stats.compilation.assets['main.js'].source()).toContain('');
        resolve(stats);
      }
    });
  });
}

describe('plugin', () => {

  it('should contain metadata', async () => {
    
    const options = {
      context: `${project2Path}/website`,
      plugins: [
        new Plugin({
          context: project2Path
        })
      ]
    };

    const stats = await run(options);
    expect(stats.compilation.errors).toHaveLength(0);
    expect(stats.compilation.assets['main.js'].source()).toMatchSnapshot();

  });

  it('should contain transformed metadata', async () => {
    
    const options = {
      context: `${project2Path}/website`,
      plugins: [
        new Plugin({
          context: project2Path,
          transform: () => ({foo: 'bar'})
        })
      ]
    };

    const stats = await run(options);
    expect(stats.compilation.errors).toHaveLength(0);
    expect(stats.compilation.assets['main.js'].source()).toMatchSnapshot();

  });

  it('should contain stringified metadata', async () => {
    
    const options = {
      context: `${project2Path}/website`,
      plugins: [
        new Plugin({
          context: project2Path,
          stringify: () => `export default {load: () => ({foo: 'bar'})}`
        })
      ]
    };

    const stats = await run(options);
    expect(stats.compilation.errors).toHaveLength(0);
    expect(stats.compilation.assets['main.js'].source()).toMatchSnapshot();

  });

  it('should call transform with the loader context', async () => {
    
    const options = {
      context: `${project2Path}/website`,
      plugins: [
        new Plugin({
          context: project2Path,
          transform() {
            expect(this).toHaveProperty('addDependency', expect.any(Function));
            return {};
          }
        })
      ]
    };

    await run(options);

  });

  it('should use a different module name', async () => {
    
    const options = {
      context: `${project3Path}/website`,
      plugins: [
        new Plugin({
          module: 'metadata',
          context: project3Path
        })
      ]
    };

    await run(options);

  });

});
