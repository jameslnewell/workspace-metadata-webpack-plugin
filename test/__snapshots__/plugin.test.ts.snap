// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin should contain metadata 1`] = `"!function(e){var t={};function r(n){if(t[n])return t[n].exports;var o=t[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,r),o.l=!0,o.exports}r.m=e,r.c=t,r.d=function(e,t,n){r.o(e,t)||Object.defineProperty(e,t,{enumerable:!0,get:n})},r.r=function(e){\\"undefined\\"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:\\"Module\\"}),Object.defineProperty(e,\\"__esModule\\",{value:!0})},r.t=function(e,t){if(1&t&&(e=r(e)),8&t)return e;if(4&t&&\\"object\\"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(r.r(n),Object.defineProperty(n,\\"default\\",{enumerable:!0,value:e}),2&t&&\\"string\\"!=typeof e)for(var o in e)r.d(n,o,function(t){return e[t]}.bind(null,o));return n},r.n=function(e){var t=e&&e.__esModule?function(){return e.default}:function(){return e};return r.d(t,\\"a\\",t),t},r.o=function(e,t){return Object.prototype.hasOwnProperty.call(e,t)},r.p=\\"\\",r(r.s=0)}([function(e,t,r){\\"use strict\\";r.r(t);var n={};r.r(n),r.d(n,\\"default\\",function(){return o});var o={root:{path:'\\"../test/__fixtures__/project-2\\"',json:{name:\\"project-2\\",workspaces:[\\"components/{atom,molecule}/*\\",\\"website\\"]}},workspaces:[{path:'\\"../test/__fixtures__/project-2/website\\"',json:{name:\\"@ui/website\\",private:!0}},{path:'\\"../test/__fixtures__/project-2/components/atom/button\\"',json:{name:\\"@ui/button\\"}},{path:'\\"../test/__fixtures__/project-2/components/molecule/feature-panel\\"',json:{name:\\"@ui/feature-panel\\"}}]};console.log(n)}]);"`;

exports[`plugin should contain stringified metadata 1`] = `"!function(e){var r={};function t(n){if(r[n])return r[n].exports;var o=r[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,t),o.l=!0,o.exports}t.m=e,t.c=r,t.d=function(e,r,n){t.o(e,r)||Object.defineProperty(e,r,{enumerable:!0,get:n})},t.r=function(e){\\"undefined\\"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:\\"Module\\"}),Object.defineProperty(e,\\"__esModule\\",{value:!0})},t.t=function(e,r){if(1&r&&(e=t(e)),8&r)return e;if(4&r&&\\"object\\"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(t.r(n),Object.defineProperty(n,\\"default\\",{enumerable:!0,value:e}),2&r&&\\"string\\"!=typeof e)for(var o in e)t.d(n,o,function(r){return e[r]}.bind(null,o));return n},t.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(r,\\"a\\",r),r},t.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},t.p=\\"\\",t(t.s=0)}([function(e,r,t){\\"use strict\\";t.r(r);var n={};t.r(n),t.d(n,\\"default\\",function(){return o});var o={load:()=>({foo:\\"bar\\"})};console.log(n)}]);"`;

exports[`plugin should contain transformed metadata 1`] = `"!function(e){var r={};function t(n){if(r[n])return r[n].exports;var o=r[n]={i:n,l:!1,exports:{}};return e[n].call(o.exports,o,o.exports,t),o.l=!0,o.exports}t.m=e,t.c=r,t.d=function(e,r,n){t.o(e,r)||Object.defineProperty(e,r,{enumerable:!0,get:n})},t.r=function(e){\\"undefined\\"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:\\"Module\\"}),Object.defineProperty(e,\\"__esModule\\",{value:!0})},t.t=function(e,r){if(1&r&&(e=t(e)),8&r)return e;if(4&r&&\\"object\\"==typeof e&&e&&e.__esModule)return e;var n=Object.create(null);if(t.r(n),Object.defineProperty(n,\\"default\\",{enumerable:!0,value:e}),2&r&&\\"string\\"!=typeof e)for(var o in e)t.d(n,o,function(r){return e[r]}.bind(null,o));return n},t.n=function(e){var r=e&&e.__esModule?function(){return e.default}:function(){return e};return t.d(r,\\"a\\",r),r},t.o=function(e,r){return Object.prototype.hasOwnProperty.call(e,r)},t.p=\\"\\",t(t.s=0)}([function(e,r,t){\\"use strict\\";t.r(r);var n={};t.r(n),t.d(n,\\"default\\",function(){return o});var o={foo:\\"bar\\"};console.log(n)}]);"`;
